File: go/ArrayIsGood/ArrayIsGood.go
package main

import "fmt"

// Time -> 0(N) Space -> 0(N)
func isGood(nums []int) bool {
	base := len(nums) - 1
	baseMap := make([]int, base+1)
	for _, num := range nums {
		if num <= 0 || num > base {
			return false
		}

		baseMap[num]++

		if baseMap[num] > 2 || (baseMap[num] == 2 && num != base) {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(isGood([]int{1, 2, 3, 3}))
	fmt.Println(isGood([]int{0, 1, 2, 3}))
}

File: python/arrayIsGood.py
# Time -> 0(N) Space -> 0(N)
def isGood(nums: list[int]) -> bool:
    base = len(nums) - 1
    baseMap = [0] * (base + 1)
    for num in nums:
        if num <= 0 or num > base:
            return False
        baseMap[num] += 1

        if baseMap[num] > 2 or (baseMap[num] == 2 and num != base):
            return False

    return True


print(isGood([1, 2, 3, 3]))
print(isGood([0, 1, 2, 3]))
